pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -158,6 +158,51 @@ protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutab
     }
   }
 
+  /**
+   * <li> When the replacing segment and current segment are of {@link LLCSegmentName} then the PK should resolve to
+   * row in segment with higher sequence id.
+   * <li> When the replacing segment and current segment are of {@link UploadedRealtimeSegmentName} then the PK
+   * should resolve to row in segment with higher sequence id, creation time.
+   * <li> When either is of type {@link UploadedRealtimeSegmentName} then resolve on creation time, if same(rare
+   * scenario) then give preference to uploaded time
+   *
```

Comment:
afaik, we can create segments using `SegmentIndexCreationDriverImpl` and creationTime is always populated. The segment name format may be unknown but the segment is still uploaded and can be for say, data correction/addition.

So we can rely on creation time as first comparison criterion followed by below criteria:
LLC, Unknown: creation time check or prefer unknown as this is an uploaded segment too (deviates from current behaviour)
Uploaded, Unknown: creation time check or prefer known name format
Unknown, Unknown: creation time check or false

Reply:
sgtm, basically:
1. if both LLC, compare seqId
2. if both Uploaded, compare ctime, then seqId (btw, I refined this point based on one unit test in this PR. I commented on that unit test as well.)
3. otherwise, simply check ctimes. If ctimes are same, favor uploaded segment

Reply:
> if both Uploaded, compare ctime, then seqId (btw, I refined this point based on one unit test in this PR. I commented on that unit test as well.)

This makes sense, thinking from the backfill perspective(corrective/bootstrap backfill). Batch jobs run at a time T and the segments for that time T will be generated sequentially.

The naming convention java docs can call this out. Let me update the UTs accordingly.